spec/controllers/course/survey/surveys_controller_spec.rb

```diff
@@ -7,14 +7,195 @@
   with_tenant(:instance) do
     let(:admin) { create(:administrator) }
     let!(:course) { create(:course, creator: admin) }
+    let(:student) { create(:course_student, course: course) }
     let!(:survey) do
       create(:survey, *survey_traits,
              course: course, section_count: 2, section_traits: [:with_all_question_types])
     end
     let(:survey_traits) { nil }
+    let(:survey_stub) do
+      stub = create(:survey, course: course)
+      allow(stub).to receive(:save).and_return(false)
+      allow(stub).to receive(:update_attributes).and_return(false)
+      allow(stub).to receive(:destroy).and_return(false)
+      stub
+    end
+    let(:json_response) { JSON.parse(response.body) }
 
     before { sign_in(user) }
 
+    describe '#index' do
+      let!(:published_survey) { create(:survey, :published, course: course) }
+
+      context 'when html page is requested' do
+        let(:user) { student.user }
+        subject { get :index, course_id: course.id }
+
+        it { is_expected.to render_template('index') }
+      end
+
+      context 'when json data is requested' do
+        render_views
+        subject { get :index, format: :json, course_id: course.id }
+        before { subject }
+
+        context 'when user is staff' do
+          let(:user) { admin }
+
+          it 'sees all surveys' do
+            expect(json_response['surveys'].length).to eq(2)
+          end
+        end
+
+        context 'when user is student' do
+          let(:user) { student.user }
+
+          it 'sees only published surveys' do
+            expect(json_response['surveys'].length).to eq(1)
+          end
+        end
+      end
+    end
+
+    describe '#create' do
+      let(:user) { admin }
+
+      subject do
+        post :create, format: :json, course_id: course.id, id: survey.id,
+                      survey: attributes_for(:survey)
+      end
+
+      it 'creates a survey' do
+        expect { subject }.to change { course.surveys.count }.by(1)
+      end
+
+      context 'when saving fails' do
+        before do
+          controller.instance_variable_set(:@survey, survey_stub)
+          subject
+        end
+
+        it { is_expected.to have_http_status(:bad_request) }
+      end
+    end
+
+    describe '#show' do
+      let(:survey_traits) { :published }
+      let(:user) { student.user }
+
+      context 'when html page is requested' do
+        subject { get :show, course_id: course.id, id: survey.id }
+
+        it { is_expected.to render_template('index') }
+      end
+
+      context 'when html page is requested' do
+        subject { get :show, format: :json, course_id: course.id, id: survey.id }
+
+        it { is_expected.to render_template('_survey_with_questions') }
+      end
+    end
+
+    describe '#update' do
+      let(:user) { admin }
+      let(:survey_params) { {} }
```

Comment:
You can use `attributes_for(:xx_factory)` from factory girl to generate the params, so that all types of params can be tested.